pkg/policy/l4.go

```diff
@@ -630,55 +598,42 @@ func (l4 *L4Filter) toMapState(p *EndpointPolicy, features policyFeatures, chang
 			KeyForDirection(direction).WithPortProtoPrefix(proto, mp.port, uint8(bits.LeadingZeros16(^mp.mask))))
 	}
 
-	// find the L7 rules for the wildcard entry, if any
-	var wildcardRule *PerSelectorPolicy
-	if l4.wildcard != nil {
-		wildcardRule = l4.PerSelectorPolicies[l4.wildcard]
-	}
-
-	isL4Wildcard := (l4.Port != 0 || l4.PortName != "") && l4.wildcard != nil
-	for cs, currentRule := range l4.PerSelectorPolicies {
-		// have wildcard and this is an L3L4 key?
-		isL3L4withWildcardPresent := isL4Wildcard && cs != l4.wildcard
-
-		if isL3L4withWildcardPresent && wildcardRule.covers(currentRule) {
-			logger.WithField(logfields.EndpointSelector, cs).Debug("ToMapState: Skipping L3/L4 key due to existing L4-only key")
-			continue
-		}
-
-		isDenyRule := currentRule != nil && currentRule.IsDeny
-		isRedirect := currentRule.IsRedirect()
-		listener := currentRule.GetListener()
-		priority := currentRule.GetPriority()
-		if !isDenyRule && isL3L4withWildcardPresent && !isRedirect {
-			// Inherit the redirect status from the wildcard rule.
-			// This is now needed as 'covers()' can pass non-redirect L3L4 rules
-			// that must inherit the redirect status from the L4-only (== L3-wildcard)
-			// rule due to auth type on the L3L4 rule being different than in the
-			// L4-only rule.
-			isRedirect = wildcardRule.IsRedirect()
-			listener = wildcardRule.GetListener()
-			priority = wildcardRule.GetPriority()
-		}
-		authReq := currentRule.getAuthRequirement()
+	// makeEntry creates a mapStateEntry for the given selector and policy
+	makeEntry := func(cs CachedSelector, currentRule *PerSelectorPolicy) *mapStateEntry {
 		var proxyPort uint16
-		if isRedirect {
+		if currentRule.IsRedirect() {
 			var err error
-			proxyPort, err = p.LookupRedirectPort(l4.Ingress, string(l4.Protocol), port, listener)
+			proxyPort, err = p.LookupRedirectPort(l4.Ingress, string(l4.Protocol), port, currentRule.GetListener())
 			if err != nil {
 				// Skip unrealized redirects; this happens routineously just
 				// before new redirects are realized. Once created, we are called
 				// again.
 				logger.WithError(err).WithField(logfields.EndpointSelector, cs).Debugf("Skipping unrealized redirect")
-				continue
+				return nil
 			}
 		}
-		entry := newMapStateEntry(l4.RuleOrigin[cs], proxyPort, priority, isDenyRule, authReq)
 
-		if cs.IsWildcard() {
+		ms := newMapStateEntry(
+			l4.RuleOrigin[cs],
+			proxyPort,
+			currentRule.GetPriority(),
+			currentRule.getDeny(),
+			currentRule.getAuthRequirement(),
+		)
+		return &ms
```

Comment:
Taking the address here increases the number of heap allocations, as show when running the `BenchmarkRegenerateCIDRDenyPolicyRules` benchmark. How about returning a value instead and using the `MapStateEntry.Invalid` bit to differentiate between a valid and invalid entry?

Reply:
Done. I verified the benchmarks are now unchanged.

I got scared when one benchmark got 20% slower. But it was my antivirus messing things up.